### PR TITLE
Orders: Display split view on large screens

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #6381 

### Description

This PR updates the Orders tab to show split view on large screens to take advantage of the available real estate.

The solution includes:
- Add a custom split view controller to be reused within the app. This split view controller uses the modern API to configure a double-column layout with the primary column being displayed when collapsed.
- Update the root view of the Orders tab to a new view controller that wraps the old order root within a split view controller. This is needed because the modern split view controller (initiated with `style`) doesn't work properly when added directly to a tab view controller.
- Extract `GhostTableViewController` to be a reusable type that displays ghost animation over a table view.
- Set up the orders split view controller with the order root and a ghost table view controller.
- Update the order list to automatically update the split view's detail view after loading orders
- Replace the order detail storyboard with a nib file and update the order detail view controller's initializer for easy setup.
- Additionally: update constraints on order root to make the view expand edge-to-edge in landscape mode.
- Also: updated UI tests to reflect the change to the Orders tab

### Testing instructions

On iPad (or large iPhones in landscape mode):
- Log in to a valid account. Notice that the details for the Orders tab in the tab bar are correct.
- Select the Orders tab. Notice that the list is displayed as a split view, with a placeholder view for the 2nd column initially.
- After a while, notice that details for the first item in the list are displayed on the 2nd column.
- Select another item, notice that the correct details are displayed on the 2nd column.

On iPhone (with small size):
- Log in to a valid account. Notice that the details for the Orders tab in the tab bar are correct.
- Select the Orders tab. Notice that the order list is displayed correctly.
- Select any order, notice that the correct details are displayed.

### Screenshots

On iPad:
https://user-images.githubusercontent.com/5533851/157225577-8f6aa477-867a-452c-90d6-665449408108.mp4

On iPhone:
https://user-images.githubusercontent.com/5533851/157227051-67d31080-cb80-4693-b886-c491c27bba19.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: RELEASE-NOTES.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 8.8
 -----
+- [*] Show split view for Orders tab on large screens. [https://github.com/woocommerce/woocommerce-ios/pull/6387]
```
Comment (pmusolino) [inline RELEASE-NOTES.txt]: small nit: since this functionality has a strong impact on Orders and on our users and should be tested deeply, maybe it deserves two stars.

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift [hunk 2]
```diff
@@ -448,7 +448,7 @@ private extension MainTabBarController {
     }
 
     func createOrdersViewController(siteID: Int64) -> UIViewController {
-        OrdersRootViewController(siteID: siteID)
+        OrdersSplitViewWrapperController(siteID: siteID)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: I wasn't able to test this change with Push Notifications, but it's something we should consider before merging this.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift [hunk 3]
```diff
@@ -48,22 +52,18 @@ final class OrderDetailsViewController: UIViewController {
 
     /// Order to be rendered!
     ///
-    var viewModel: OrderDetailsViewModel! {
-        didSet {
-            reloadTableViewSectionsAndData()
-        }
-    }
+    private let viewModel: OrderDetailsViewModel
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:51]: Nice refactoring!

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift [hunk 4]
```diff
@@ -93,6 +93,13 @@ final class OrderDetailsViewController: UIViewController {
         }
     }
 
+    override func viewWillDisappear(_ animated: Bool) {
+        // scroll table view to top so that the next time the split view
+        // shows the detail screen, the table view content is at the top position
+        tableView.contentOffset = .zero
+        super.viewWillDisappear(animated)
+    }
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift]: If I understood the meaning of what you want to achieve here, I believe there is an error and probably bad behavior for the end user.
In theory, after selecting a new order, I should always see the beginning of that order. Instead, the offset is set to the same offset as the previous order. It could be ok for the user, but I believe that setting the `contentOffset` to zero here is creating the behavior that you can see below in the video. Basically, after switching tab, the view will scroll on top with some delay, but there is also a margin on top (it's not really offset zero).
Plus, you can notice that the selected order in the list, is no more selected after switching to a different tab.

https://user-images.githubusercontent.com/495617/157273616-255c0583-299a-4bbf-904d-8f1038a07f40.mp4

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift [hunk 5]
```diff
@@ -123,6 +140,7 @@ private extension OrderDetailsViewController {
         tableView.rowHeight = UITableView.automaticDimension
 
         tableView.dataSource = viewModel.dataSource
+        tableView.accessibilityIdentifier = "order-details-table-view"
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift:124]: 💯

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift [hunk 6]
```diff
@@ -350,7 +358,12 @@ extension OrderListViewController: SyncingCoordinatorDelegate {
         // Configure header container view
         let headerContainer = UIView(frame: CGRect(x: 0, y: 0, width: Int(tableView.frame.width), height: 0))
         headerContainer.addSubview(topBannerView)
-        headerContainer.pinSubviewToSafeArea(topBannerView)
+        NSLayoutConstraint.activate([
+            headerContainer.leadingAnchor.constraint(equalTo: topBannerView.leadingAnchor),
+            headerContainer.trailingAnchor.constraint(equalTo: topBannerView.trailingAnchor),
+            headerContainer.safeAreaLayoutGuide.topAnchor.constraint(equalTo: topBannerView.topAnchor),
+            headerContainer.safeAreaLayoutGuide.bottomAnchor.constraint(equalTo: topBannerView.bottomAnchor),
+        ])
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: Q: It's not clear to me why this change. Why we don't want to maintain leading and trailing anchors to the safe area?

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift [hunk 7]
```diff
@@ -394,6 +407,25 @@ extension OrderListViewController {
...
+        let secondaryColumnNavigationController = splitViewController?.viewController(for: .secondary) as? UINavigationController
+        let displayedOrderDetail = secondaryColumnNavigationController?.viewControllers.first as? OrderDetailsViewController
+
+        guard displayedOrderDetail == nil,
+              !dataSource.isEmpty,
+              splitViewController?.isCollapsed == false else {
+            return
+        }
+        let firstIndexPath = IndexPath(row: 0, section: 0)
+        tableView.selectRow(at: firstIndexPath, animated: false, scrollPosition: .top)
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: nit: maybe this is superfluous since we are already calling below `tableView(tableView, didSelectRowAt: firstIndexPath)`?

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift [hunk 8]
```diff
@@ -515,7 +547,9 @@ private extension OrderListViewController {
 extension OrderListViewController: UITableViewDelegate {
 
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
-        tableView.deselectRow(at: indexPath, animated: true)
+        if splitViewController?.isCollapsed == true {
+            tableView.deselectRow(at: indexPath, animated: true)
+        }
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: I noticed that the cell remain selected after rotating back in portrait:

File: WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift [hunk 9]
```diff
@@ -61,8 +68,9 @@ final class OrdersRootViewController: UIViewController {
         self.siteID = siteID
         super.init(nibName: Self.nibName, bundle: nil)
 
+        // workaround to get rid of the extra space at the bottom when embedded in split view
+        extendedLayoutIncludesOpaqueBars = true
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift]: 🤭 I think optimizing for big screens will give us a big headache. It is full of small details that are easy to forget.

File: WooCommerce/UITestsFoundation/Screens/Orders/OrdersScreen.swift [hunk 10]
```diff
@@ -48,10 +48,11 @@ public final class OrdersScreen: ScreenObject {
...
     public func verifyOrdersList(orders: [OrderData]) throws -> Self {
-        XCTAssertEqual(orders.count, app.tables.cells.count, "Expecting '\(orders.count)' orders, got '\(app.tables.cells.count)' instead!")
-        app.assertTextVisibilityCount(textToFind: String(orders[0].id))
-        app.assertTextVisibilityCount(textToFind: String(orders[0].total))
-        app.assertLabelContains(firstSubstring: String(orders[0].id), secondSubstring: orders[0].billing.first_name)
+        let ordersTableView = app.tables.matching(identifier: "orders-table-view")
+        XCTAssertEqual(orders.count, ordersTableView.cells.count, "Expecting '\(orders.count)' orders, got '\(app.tables.cells.count)' instead!")
+        ordersTableView.element.assertTextVisibilityCount(textToFind: String(orders[0].id))
+        ordersTableView.element.assertTextVisibilityCount(textToFind: String(orders[0].total))
+        ordersTableView.element.assertLabelContains(firstSubstring: String(orders[0].id), secondSubstring: orders[0].billing.first_name)
```
Comment (pmusolino) [inline WooCommerce/UITestsFoundation/Screens/Orders/OrdersScreen.swift:55]: Thanks for updating our UI Tests!

**Review by pmusolino — CHANGES_REQUESTED** [review]
Nice work @itsmeichigo! This is the first for a big refactoring of the app for supporting large screens and given the changes of this PR... this scares me a little bit 🤣 

By the way, I left some comments and I found some issues that should be addressed before merging the PR 👍

File: RELEASE-NOTES.txt (hunk 1 above)
Comment (itsmeichigo) [reply RELEASE-NOTES.txt]: Good point! Updated in a5fbc69d51d63f046f0a375dbc416415dc6d521b.

File: WooCommerce/Classes/ViewRelated/MainTabBarController.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/MainTabBarController.swift]: Thank you for reminding me, presenting order details indeed has some issue when the split view is introduced. This has been fixed in 7a405044e4eb28acc074248b94f9db476a0fda98.

File: WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Details/OrderDetailsViewController.swift]: I was reusing the order details screen for every selected order, so I wanted to scroll the details view to the top every time a new item is selected, so that it looks better in collapsed mode.

However, I haven't yet found a good way to scroll the table view to top successfully, so I switched to creating a new details view for every item in 47d7489ea650f57e5443c9821b458e0e71efa684. This will keep the old behavior that we had with the navigation stack.

> Plus, you can notice that the selected order in the list, is no more selected after switching to a different tab.

This is actually a different issue was handled in 3132df8a4de80b33efac478d545ba73d14c9224f and 680648fa38288ccd0a0d4be490858db27b4fe618.

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift (hunk 6 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: I want the Order list to extend to all edges - this should be safe to do when all its subviews are anchored to safe area. I fixed the issue with the layout of the footer spinner view as well as the filters bar that you captured above in 09895c0523eb83d54bcff15cf709c7f49b84658a.

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift (hunk 7 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: The `selectRow` method only highlight the specified row and doesn't trigger the delegate method, therefore I need both methods to both highlight the row and trigger the selection.

File: WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift (hunk 8 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift]: This has been fixed in 3132df8a4de80b33efac478d545ba73d14c9224f by updating the highlight state on `viewDidLayoutSubview`.

File: WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift (hunk 9 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift]: I believe we use opaque bars somewhere in the app (the navigation bars maybe), which causes the need for this. The default behavior for bars since iOS 7 is translucent, and this property only needs to be set to true if we use opaque bars.

**Comment by itsmeichigo** [comment]
Thank you @pmusolino for the thorough review! I've added a few more changes to fix issues with the layout and handling notifications. The PR is ready for another review 🙇‍♀️

**Comment by itsmeichigo** [comment]
> 

Regarding the issue with the color for the selected row, I'd need an opinion from @woocommerce/mobile-designers. By default when a row is selected, its separators are hidden. We may need to add custom views for the separators / or update the highlight color to something that stands out more.

File: WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift (hunk 9 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift]: I can confirm that this only happens with views embedded in a navigation controller that is added to a split view. Since we're using opaque navigation bars, `extendedLayoutIncludesOpaqueBars` needs to be true to get rid of the extra space at the bottom. ~Removed redundant settings for navigation controllers in d9b192ab8a0de2e463ca02b0c7b051507d1716e0.~ (reverted the change because it doesn't work on small screens 😓)

Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/OrdersRootViewController.swift]: Come to think of it, maybe we should make sure that all child view controllers of our custom navigation bar has `extendedLayoutIncludesOpaqueBars` set to true. Updated this in 8a6cc081f63f95531de7edc63466ae8b792924b7.

**Comment by joe-keenan** [comment]
I’ve looked around at different approaches to this, and I think the simplest solution, for now, is to not hide the separators.

This would ideally apply to both tap and selected states – because at the moment, if you tap on a single row on a grey background, the whole row looks hidden. If we keep the separators, then the row will still look like a row, even though the background colours are the same.

This will have the same effect on this split-screen view.

Current on-tap look:
image

Revised look with separators
image


Does this sound ok?

**Review by pmusolino — COMMENTED** [review]
I re-tried this PR, but while I was testing it on an iPad Pro (12.9 inch) 5th generation with iOS 15.2, I encountered a crash.
This is what I did:
1. I moved to Orders, in landscape mode.
2. I loaded all the orders (paginated). 
3. I returned to top, and I did more than a pull to refresh.
4. Crash 💥 

The state of the view became strange: the data on the left were no longer displayed, while on the right the ghost view disappear displaying one order.
Simulator Screen Shot - iPad Pro (12 9-inch) (5th generation) - 2022-03-10 at 11 39 53

At the same time, the crash happened.
Schermata 2022-03-10 alle 11 40 12

**Comment by itsmeichigo** [comment]
Hi @pmusolino - can you reproduce the issue repeatedly? I'm wondering if this is an existing issue of the order list, because I haven't touched the data part of the list.

**Comment by pmusolino** [comment]
> Hi @pmusolino - can you reproduce the issue repeatedly? I'm wondering if this is an existing issue of the order list, because I haven't touched the data part of the list.

Yes, I'm able to reproduce it, but just on this PR. If I pull `trunk`, I'm not able to replicate the crash 🤔

**Comment by itsmeichigo** [comment]
Hi @pmusolino! The crash seems to happen when we try to select a table view row when the table view is still loading from the snapshot updates. It might be safer to only update the selection state right before displaying the cells, so I move the highlight logic to the delegate method `willDisplayCell` in cb8c5b23a0b4bb4e5cee35f21e03ba5d39273c83. I believe it's safer there since we're dealing with solely the main thread in this case.

I've also added separators to the selected background view of order cells like @joe-keenan suggested:

Simulator Screen Shot - iPhone 13 Pro Max - 2022-03-11 at 13 14 03

I've also added a PR to hide the changes in this PR under a feature flag: #6415. Please feel free to test that PR first!

File: WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift [hunk 11]
```diff
@@ -122,8 +122,35 @@ private extension OrderTableViewCell {
...
+
+            let bottomSeparatorView = UIView()
+            bottomSeparatorView.backgroundColor = .border
+            bottomSeparatorView.translatesAutoresizingMaskIntoConstraints = false
+            view.addSubview(bottomSeparatorView)
+            NSLayoutConstraint.activate([
+                view.bottomAnchor.constraint(equalTo: bottomSeparatorView.bottomAnchor, constant: separatorHeight/2),
+                view.leadingAnchor.constraint(equalTo: bottomSeparatorView.leadingAnchor),
+                view.trailingAnchor.constraint(equalTo: bottomSeparatorView.trailingAnchor),
+                bottomSeparatorView.heightAnchor.constraint(equalToConstant: separatorHeight/2)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift:149]: I had to reduce the bottom separator height in half - this is hacky but it helps with making it look equal with other separators. Please let me know if there's a better solution!

**Review by pmusolino — APPROVED** [review]
Thanks for fixing all the issues @itsmeichigo 🚀
